server/base/src/main/java/org/apache/accumulo/server/conf/store/impl/ZooPropLoader.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.accumulo.server.conf.store.impl;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+
+import org.apache.accumulo.fate.zookeeper.ZooReaderWriter;
+import org.apache.accumulo.server.conf.codec.VersionedPropCodec;
+import org.apache.accumulo.server.conf.codec.VersionedProperties;
+import org.apache.accumulo.server.conf.store.PropCacheId;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.data.Stat;
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.github.benmanes.caffeine.cache.CacheLoader;
+
+public class ZooPropLoader implements CacheLoader<PropCacheId,VersionedProperties> {
+
+  private static final Logger log = LoggerFactory.getLogger(ZooPropLoader.class);
+
+  private final ZooReaderWriter zrw;
+  private final VersionedPropCodec propCodec;
+  // used to set watcher, does not react to events.
+  private final PropStoreWatcher propStoreWatcher;
+  private final PropStoreMetrics metrics;
+
+  public ZooPropLoader(final ZooReaderWriter zrw, final VersionedPropCodec propCodec,
+      final PropStoreWatcher propStoreWatcher, final PropStoreMetrics metrics) {
+    this.zrw = zrw;
+    this.propCodec = propCodec;
+    this.propStoreWatcher = propStoreWatcher;
+    this.metrics = metrics;
+  }
+
+  @Override
+  public @Nullable VersionedProperties load(PropCacheId propCacheId) {
+    try {
+      log.trace("load called for {}", propCacheId);
+      metrics.updateLoadCounter();
+      Stat stat = new Stat();
+      var vProps = propCodec.fromBytes(zrw.getData(propCacheId.getPath(), propStoreWatcher, stat));
+      if (stat.getVersion() != vProps.getDataVersion()) {
```

Comment:
This code is well structured in that it has dedicated code to go to/from Zookeeper <=> VersionProperties .  Code uses w/ the VersionProperties POJO which has a version, but the code using it does not have to care how or where that is persisted in ZK.  For correctness the code that persist to ZK has to care about what the version means relative to the stat object.  To me the code that persist the POJO to from ZK should be optimized for ZK as long as that does not leak outside to other code (like push ZK types out), which I don't think it would.

Reply:
> The only place where the ZooKeeper stat and the version need to be both known is on the ZooKeeper writes / reads - and I tried to isolate that to the ZooPropLoader.

ZooPropStore seems to also use the stat.

> Code that uses versioned props does not need to worry about the version. The version only matters on store / reads and then the ZooKeeper specifics are isolated (I hope) to ZooPropLoader - so I think I was trying to achieve the separation that you are asking for - or I'm missing what you are saying?

The separation is really nice, not advocating any change to that.  Just seems nicer to me if within the ZooPropLoader and ZooPropStore code that the Stat is the authoritative source of the persisted version not the version stored in the value.  So for write to ZK

```java
    zk.put(serializeEverythingButVersion(verProps), verProps.getVersion())
``` 

and read from ZK do the following.

```java
Stat stat = new stat;
byte data = zk.read(propPath, stat)
var verProp = newVersionedProps(data, stat)
```

If within these two classes ZK stat is the authority then do not need to do the following

 * check the version in the value against the stat in ZK after every read
 * make any assumptions about what the next version in ZK will be